quickwit/quickwit-search/src/collector.rs

```diff
@@ -695,13 +699,149 @@ pub(crate) fn make_merge_collector(
     })
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct PartialHitSortingKey {
+    score: Option<SortValue>,
+    address: GlobalDocAddress,
+    asc: bool,
+}
+
+impl Ord for PartialHitSortingKey {
+    fn cmp(&self, other: &PartialHitSortingKey) -> Ordering {
+        assert_eq!(
+            self.asc, other.asc,
+            "comparing two PartialHitSortingKey of different ordering"
+        );
+        if self.asc {
+            self.score
+                .map(Reverse)
+                .cmp(&other.score.map(Reverse))
+                .then(other.address.cmp(&self.address))
+        } else {
+            self.score
+                .cmp(&other.score)
+                .then(self.address.cmp(&other.address))
+        }
+    }
+}
+
+impl PartialOrd for PartialHitSortingKey {
+    fn partial_cmp(&self, other: &PartialHitSortingKey) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+fn sort_asc(partial_hit: &PartialHit) -> PartialHitSortingKey {
+    PartialHitSortingKey {
+        score: partial_hit.sort_value,
+        address: GlobalDocAddress::from_partial_hit(partial_hit),
+        asc: true,
+    }
+}
+fn sort_desc(partial_hit: &PartialHit) -> PartialHitSortingKey {
+    PartialHitSortingKey {
+        score: partial_hit.sort_value,
+        address: GlobalDocAddress::from_partial_hit(partial_hit),
+        asc: false,
+    }
+}
+
+/// Incrementally merge segment results.
+pub(crate) struct IncrementalCollector {
+    inner: QuickwitCollector,
+    top_k_hits: TopK<PartialHit, PartialHitSortingKey>,
+    intermediate_aggregation_results: Vec<Vec<u8>>,
+    num_hits: u64,
+    failed_splits: Vec<SplitSearchError>,
+    num_attempted_splits: u64,
+}
+
+impl IncrementalCollector {
+    /// Create a new incremental collector
+    pub(crate) fn new(inner: QuickwitCollector) -> Self {
+        let sort_key_fn = match inner.sort_by.sort_order() {
+            SortOrder::Asc => sort_asc,
+            SortOrder::Desc => sort_desc,
```

Comment:
Yeah at this point I think you would have been better off using SortOrder in the PartialHitSortingKey :D

Reply:
part of why this is needed is because I did not want TopK to have a type parameter (which is required when using closures, unless you box it), so I used function pointers directly.
Having a type parameter on TopK means IncrementalCollector would have one too, and I don't think it would be possible to name the type output of `IncrementalCollector::new` due to that anonymous closure.
Alternatively, we could have a simple `SortKeyMapper` trait or something along the line, with a single function doing the same as `sort_key_fn`, and an implementer which can be named

wdyt?

Reply:
Yes yes. I understood you wanted to avoid having a generic type. This is looking good.

My comment was more in reference to another place were you used a bool to say Ascending / Descending. Had 2 constructors for each, and an if statement. Only to eventually convert it to the `SortOrder` enum.

Reply:
(I think you already modified the code)